mongoengine/queryset/base.py

```diff
@@ -1101,6 +1088,20 @@ def order_by(self, *keys):
 
         return queryset
 
+    def clear_cls_query(self):
+        """Clear the default "_cls" query.
+
+        By default, all queries generated for documents that allow inheritance
+        include an extra "_cls" clause. In most cases this is desirable, but
+        sometimes you might achieve better performance if you clear that
+        default query.
```

Comment:
More by curiosity but could you elaborate on the use case for this? Wouldn't it be the same (and more explicit) to query on the parent class?

Reply:
Correct me if I'm wrong about any of these points, but:
* The parent class still adds an initial cls query, which sometimes isn't needed.
* I believe we're using `Parent.objects.filter(__raw__={'_cls': {'$in': ['Parent.Child1', 'Parent.Child2']}}).clear_cls_query()` to filter by specific subclasses.
* If you filter by primary key, you might not need the initial cls query.

Reply:
if you filter by PK, I would think that mongo would use only its index anyway so I don't think it will improve performance

Reply:
> Wouldn't it be the same (and more explicit) to query on the parent class?

Yes, definitely. It's 100% clearer to do `Parent.objects(...)` vs `Child.objects(...).clear_cls_query()`. That said, this method isn't really meant to be used on the child classes. It's meant to be used on the **parent** class.

Imagine having an `Animal` parent class and a bunch of subclasses: `Dog`, `Cat`, `Horse`, etc. If I want to fetch all animals belonging to a certain person, this is the query that's generated by default:
```py
In [10]: class Animal(Document):
    ...:     owner = StringField()
    ...:     meta = {'allow_inheritance': True}
    ...:
    ...: class Dog(Animal):
    ...:     pass
    ...:
    ...: class Cat(Animal):
    ...:     pass
    ...:
    ...: class Horse(Animal):
    ...:     pass
    ...:

In [11]: Animal.objects(owner='George')._query
Out[11]:
{'_cls': {'$in': ('Animal', 'Animal.Dog', 'Animal.Cat', 'Animal.Horse')},
 'owner': 'George'}
```

Now, imagine that I have an `{"owner": 1}` index on that collection and that there's 100M+ documents in it. *Most likely* MongoDB will use the right index (although I've seen it get confused at times and e.g. use a `{"_cls": 1}` index). But even then, this query will result in unnecessary filtering after an `IXSCAN`. Its performance will be suboptimal. On the other hand, `Animal.objects(owner='George').clear_cls_query()` would:
1. Use the right index w/o any redundant filtering.
1. Transmit less data over the wire because the query is smaller (this is less important, but still...).

> if you filter by PK, I would think that mongo would use only its index anyway so I don't think it will improve performance

Yep. The only (small, negligible?) improvement in this case is point (2) mentioned above (sending less data over the wire to MongoDB).

---

An alternative solution would be to not include the `"_cls": {"$in": [...]}` clause *by default* when querying the parent class. Most of the time it's redundant. This might lead to errors if you have some leftover documents in your collection with a `_cls` that isn't mapped in your code, but that's an edge case we likely shouldn't optimize for.

This would only include a `"_cls": "Parent.Subclass"` clause when querying the subclass (e.g. `Dog.objects()`).

If I were designing MongoEngine from scratch, this is likely what I'd do. But given where we are, I think `clear_cls_query()` might be an easier path forward. What do y'all think @bagerard @thomasst ?

Reply:
> An alternative solution would be to not include the `"_cls": {"$in": [...]}` clause _by default_

That's fine with me, but you might still want to have a `clear_cls_query()` clause if you're trying to optimize a query for a subclass when also filtering by a field that's only present in the subclass. Also, it's a breaking change that's a bit difficult to detect – it might break existing apps with leftover documents (as you mentioned), and you don't usually test your app for that.

Reply:
Fair points. Given these concerns, I think we should stay with the implementation that this PR offers currently.

Reply:
Thanks for the detailed explanations, it makes sense